### PR TITLE
Support to --emit-externs

### PR DESCRIPTION
**Pull request by engjefersonsantiago**
Issue #1216

### CODE REVIEW

File: backends/bmv2/extern.cpp [hunk 1]
```diff
@@ -75,9 +75,9 @@ bool Extern::preorder(const IR::Declaration_Instance* decl) {
 
 /// Custom visitor to enable traversal on other blocks
 bool Extern::preorder(const IR::PackageBlock *block) {
-    if (backend->target != Target::PORTABLE_SWITCH)
+    if (backend->target != Target::PORTABLE_SWITCH && !emitExterns)
         return false;
-
+    std::cout << "je suis ici\n";
```
Comment (mbudiu-vmw) [inline backends/bmv2/extern.cpp]: This should be removed.
You should use LOG for log messages.

File: backends/bmv2/simpleSwitch.cpp [hunk 2]
```diff
@@ -206,7 +207,25 @@ SimpleSwitch::convertExternObjects(Util::JsonArray *result,
             // Do not generate any code for this operation
         }
     } else {
-        error("Unknown extern type %1%", em->originalExternType->name);
+        if (emitExterns) {
+            ::warning("Found extern method %1%. Make sure that the BMv2 supports it",
```
Comment (mbudiu-vmw) [inline backends/bmv2/simpleSwitch.cpp]: I would think that this warning is unnecessary - if the user has passed the suitable flag to the compiler they already should know this.

File: frontends/common/options.cpp [hunk 3]
```diff
@@ -121,6 +121,9 @@ CompilerOptions::CompilerOptions() : Util::Options(defaultMessage) {
                        }
                        return true; },
                    "Choose output format for the P4Runtime API description (default is binary).");
+    registerOption("--emit-externs", nullptr,
```
Comment (mbudiu-vmw) [inline frontends/common/options.cpp]: This option does not belong to common - it should be specific to the back-end.
Also, the documentation has to be more expansive, describing what will actually be generated in the output code.

File: backends/bmv2/simpleSwitch.h [hunk 4]
```diff
@@ -46,12 +46,12 @@ class SimpleSwitch {
  public:
     void modelError(const char* format, const IR::Node* place) const;
     void convertExternObjects(Util::JsonArray *result, const P4::ExternMethod *em,
-                              const IR::MethodCallExpression *mc, const IR::StatOrDecl *s);
+                              const IR::MethodCallExpression *mc, const IR::StatOrDecl *s, const bool& emitExterns);
```
Comment (mbudiu-vmw) [inline backends/bmv2/simpleSwitch.h]: Did you run cpplint? These lines are too long.

File: backends/bmv2/simpleSwitch.cpp [hunk 5]
```diff
@@ -586,6 +606,13 @@ SimpleSwitch::convertExternInstances(const IR::Declaration *c,
         }
 
         action_profiles->append(action_profile);
+    } else {
+        if (emitExterns) {
```
Comment (mbudiu-vmw) [inline backends/bmv2/simpleSwitch.cpp]: Shouldn't something happen for this case too?

File: backends/bmv2/simpleSwitch.h (hunk 4 above)
Comment (engjefersonsantiago) [reply backends/bmv2/simpleSwitch.h]: I ran. I realized that the headers files were not include in the rule for the cpplint. My next commit will include it.

**Comment by engjefersonsantiago** [comment]
Hi @mbudiu-vmw , I think I've addressed your comments.

File: backends/bmv2/extern.cpp [hunk 6]
```diff
@@ -19,29 +19,28 @@ limitations under the License.
...
                             const IR::ExternBlock* block) {
     auto attributes = new Util::JsonArray();
-    auto paramIt = block->getConstructorParameters()->parameters.begin();
-    for (auto arg : *di->arguments) {
-        auto name = arg->toString();
-        if (arg->is<IR::Constant>()) {
-            auto cVal = arg->to<IR::Constant>();
-            if (arg->type->is<IR::Type_Bits>()) {
+    for (auto p : *block->getConstructorParameters()) {
+        auto name = p->toString();
```
Comment (mbudiu-vmw) [inline backends/bmv2/extern.cpp]: You should use `p->name` and not `p->toString` here and below.

**Pull request merged**